### PR TITLE
[Shape Inference] ScatterUpdate - add evaluate_label

### PR DESCRIPTION
**Pull request by mitruska**
### Details:
 - Add custom `evaluate_label` method for ScatterUpdate to preserve labels from updates input values
   (Default label evaluator works only when value labels are in the first input)
 
 - TypeProp tests for interval dimensions, dimension labels and value labels propagation

### Tickets:
 - 90624 (90721, 90758, 90943)

### CODE REVIEW

File: src/core/src/op/scatter_update.cpp [hunk 1]
```diff
@@ -124,3 +124,65 @@ bool op::v3::ScatterUpdate::has_evaluate() const {
...
+
+    if (in0_has_no_labels && in2_has_no_labels) {
+        return false;
+    }
+
+    HostTensorVector input_tensors(input_values.size());
+    for (size_t i = 0; i < input_values.size(); ++i) {
+        const auto& input = input_values[i];
+        if (i == data_in_idx || i == updates_in_idx) {
+            auto input_labels = input.get_tensor().get_value_label();
```
Comment (praasz) [inline src/core/src/op/scatter_update.cpp:149]: It looks like get labels and check `no label` for input 0, 2 is done twice (previous at L:138,139). Can be reduced to do this once per input?

File: src/core/src/op/scatter_update.cpp [hunk 2]
```diff
@@ -124,3 +124,65 @@ bool op::v3::ScatterUpdate::has_evaluate() const {
...
+        }
+    }
+
+    HostTensorVector output_tensors;
+    output_tensors.reserve(node->get_output_size());
+    for (size_t i = 0; i < node->get_output_size(); ++i) {
+        output_tensors.emplace_back(std::make_shared<HostTensor>(element::u64, node->get_output_partial_shape(i)));
+    }
+
+    if (node->evaluate(output_tensors, input_tensors)) {
```
Comment (praasz) [inline src/core/src/op/scatter_update.cpp:168]: evaluate which use `HostTensorVector` is  marked as deprecated. 
As this is new code consider use `ov::Tensor` and evaluate using `ov::TensordVector`

File: src/core/src/op/scatter_update.cpp [hunk 3]
```diff
@@ -124,3 +124,65 @@ bool op::v3::ScatterUpdate::has_evaluate() const {
...
+
+    HostTensorVector input_tensors(input_values.size());
+    for (size_t i = 0; i < input_values.size(); ++i) {
+        const auto& input = input_values[i];
+        if (i == data_in_idx || i == updates_in_idx) {
+            auto input_labels = input.get_tensor().get_value_label();
+            if (ov::has_no_labels(input_labels)) {
+                input_labels = std::vector<size_t>{0};
+            }
+            auto labels_constant = op::v0::Constant::create(ov::element::u64, input.get_shape(), input_labels);
```
Comment (praasz) [inline src/core/src/op/scatter_update.cpp:153]: Why not create the `HostTensor` or `ov::Tensor` directly without creating `Constant`?

File: src/core/src/op/scatter_update.cpp [hunk 4]
```diff
@@ -124,3 +124,65 @@ bool op::v3::ScatterUpdate::has_evaluate() const {
...
+        } else if (input.get_tensor().has_and_set_bound()) {
+            input_tensors[i] = input.get_tensor().get_lower_value();
+        } else {
+            return false;
+        }
+    }
+
+    HostTensorVector output_tensors;
+    output_tensors.reserve(node->get_output_size());
+    for (size_t i = 0; i < node->get_output_size(); ++i) {
```
Comment (praasz) [inline src/core/src/op/scatter_update.cpp:164]: As this operator has got only one output it the output vector can be created with one element and without loop same for `std::transform` can be removed and conversion can be done form first element only.

File: src/core/src/op/scatter_update.cpp (hunk 1 above)
Comment (mitruska) [reply src/core/src/op/scatter_update.cpp]: Done

File: src/core/src/op/scatter_update.cpp [hunk 5]
```diff
@@ -124,3 +124,65 @@ bool op::v3::ScatterUpdate::has_evaluate() const {
...
+    for (size_t i = 0; i < input_values.size(); ++i) {
+        const auto& input = input_values[i];
+        if (i == data_in_idx || i == updates_in_idx) {
+            auto input_labels = input.get_tensor().get_value_label();
+            if (ov::has_no_labels(input_labels)) {
+                input_labels = std::vector<size_t>{0};
+            }
+            auto labels_constant = op::v0::Constant::create(ov::element::u64, input.get_shape(), input_labels);
+            input_tensors[i] = std::make_shared<HostTensor>(labels_constant);
+        } else if (input.get_tensor().has_and_set_bound()) {
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/src/op/scatter_update.cpp]: Ok, removed.

File: src/core/src/op/scatter_update.cpp (hunk 4 above)
Comment (mitruska) [reply src/core/src/op/scatter_update.cpp]: Done

File: src/core/src/op/scatter_update.cpp (hunk 2 above)
Comment (mitruska) [reply src/core/src/op/scatter_update.cpp:170]: (Change in progress) I considered it too, I will try to update `HostTensor` to `ov::Tensor`.

File: src/core/src/op/scatter_update.cpp (hunk 3 above)
Comment (mitruska) [reply src/core/src/op/scatter_update.cpp]: (Change in progress) I considered it too, I will try to update `HostTensor` to `ov::Tensor`.

Comment (mitruska) [reply src/core/src/op/scatter_update.cpp]: Done

File: src/core/src/op/scatter_update.cpp (hunk 2 above)
Comment (mitruska) [reply src/core/src/op/scatter_update.cpp:163]: Done

File: src/core/src/op/scatter_update.cpp [hunk 6]
```diff
@@ -124,3 +124,65 @@ bool op::v3::ScatterUpdate::has_evaluate() const {
     }
     return false;
 }
+
+namespace {
+OPENVINO_SUPPRESS_DEPRECATED_START
+bool scatter_label_evaluator(const Node* node, TensorLabelVector& output_labels) {
+    const auto& input_values = node->input_values();
+    if (input_values.size() < 4) {
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/src/op/scatter_update.cpp]: Removed

File: src/core/src/op/scatter_update.cpp [hunk 7]
```diff
@@ -122,3 +122,59 @@ bool op::v3::ScatterUpdate::has_evaluate() const {
...
+        return false;
+    }
+
+    constexpr auto element_type = (sizeof(size_t) == 8) ? element::u64 : element::u32;
+    std::vector<ov::runtime::Tensor> input_tensors;
+    input_tensors.reserve(input_values.size());
+
+    auto make_input_label = [&](const Output<Node>& input, TensorLabel& labels) {
+        input_tensors.emplace_back(element_type, input.get_shape());
+        labels.resize(shape_size(input.get_shape()));
```
Comment (praasz) [inline src/core/src/op/scatter_update.cpp:145]: Detail or next time
```suggestion
        labels.resize(shape_size(input.get_shape()), ov::no_label);
```
To be sure if new elements have got correct value.

**Review by praasz — APPROVED** [review]

File: src/core/src/op/scatter_update.cpp [hunk 8]
```diff
@@ -122,3 +122,59 @@ bool op::v3::ScatterUpdate::has_evaluate() const {
     }
     return false;
 }
+
+namespace {
+bool scatter_label_evaluator(const Node* node, TensorLabelVector& output_labels) {
```
Comment (jane-intel) [inline src/core/src/op/scatter_update.cpp:127]: Is there a way to extend default label evaluator for that case too?

**Comment by jane-intel** [comment]
Agreed that @mitruska will generalize default label evaluator in the future commits

**Pull request merged**